core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -325,7 +341,15 @@ private ManifestFile filterManifest(Schema tableSchema, ManifestFile manifest) {
     }
   }
 
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
   private boolean canContainDeletedFiles(ManifestFile manifest) {
+    boolean manifestReferencedInDelete = manifestsWithDeletedFiles.contains(manifest.path());
+    if (manifest.content() == ManifestContent.DELETES && allDeletesReferenceManifests) {
```

Comment:
The reason for this condition on the manifest being a delete manifest is that there is still a remaining case that I'm seeing where we cannot use the `manifestsWithDeletedFiles` set as an absolute source of truth. The case showed up in [this test](https://github.com/apache/iceberg/blob/main/core/src/test/java/org/apache/iceberg/TestSnapshotManager.java#L58).

When performing the cherry pick operation with a dynamic overwrite, all the deleted operations (the deletion of FILE_A) do reference manifests. 

However, the manifest that actually gets passed for filtering is the manifest produced for the appending of file A. Even though it's not in the manifestsWtihDeletedFiles set, we actually do need to rewrite in this case because the cherry pick is replacing the file with FILE_A_REPLACEMENT. 

In essence, we relied on the deleted file partitions for this check previously and it seems like we still need to.

This implementation does avoid the false positive reads for delete manifests which will be important for sync maintenance as well as also efficiently identifies for both data/delete manifests which are referenced, and that seems to bring a benefit.

Reply:
Thought about this a bit more, and here is my simplified, more general explanation:

The above happens when the referenced manifest for a delete is different than the actual manifest which needs to be rewritten. The cherrypick overwrite case is one example since the referenced manifest is read from the snapshot being cherry-picked, but the actual manifest that needs to be rewritten is the manifest from the original append of FILE_A. Worth thinking through if there's more such cases and if it's possible to make this more generalizable. 

In the end, the main aspect to prevent is cases where the logic says the manifest doesn't need to be read but it actually does. Then we aim to minimize false positives on top of that.

Reply:
Note: We could also simplify all this to only track delete manifests which have entries which are deleted. But I think that loses out on the ability for the true positive case where a data manifest has a deleted entry and we can quickly return true and skip the manifest read. cc @aokolnychyi let me know what you think.

Reply:
This makes it a bit more tricky, though. Let's think more about it. Any thoughts, @amogh-jahagirdar?

Reply:
Updated so that `MergingSnapshotProducer` implementations can surface if it's possible to use the referenced manifest. I did it a bit differently than "filter manifests prior to passing them to ManifestFIlterManager". The reason is there's a few cases that need to be upleveled from `ManifestFIlterManager" (e.g. are there live files?) and it actually ends up making the diff quite a bit bigger. It's possible, it just requires refactoring quite a lot and in the end feels less readable. Instead manifest filter manager exposes a `useReferencedMnaifests` method which will get set accordingly in `MergingSnapshotProducer`. I also don't know if we need to pass in `TableMetadata` because for detecting if there was a concurrent write, I think we just need to compare the parent snapshot ID to the starting snapshot ID and make sure it's the same.

Reply:
I should write a unit test though for the replace + concurrent manifest rewrite